File: sdk/python/rollout_sdk/evaluator.py
"""Local flag evaluation engine mirroring the Go implementation."""

from __future__ import annotations

import hashlib
import re
import struct
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Tuple

from .models import (
    Clause,
    EvalContext,
    EvalResult,
    Flag,
    FlagEnvironment,
    RolloutConfig,
    TargetingRule,
)


# ─── Hash Bucket (mirrors Go hashBucket) ─────────────────────────────────────


def hash_bucket(key: str, seed: int) -> int:
    """
    Deterministic bucket in [0, 10000) for sticky bucketing.

    Mirrors the Go implementation:
      - Write seed as 8-byte little-endian uint64
      - Write key as UTF-8 bytes
      - SHA-256, take first 4 bytes as big-endian uint32, mod 10000
    """
    seed_bytes = struct.pack("<Q", seed & 0xFFFFFFFFFFFFFFFF)
    key_bytes = key.encode("utf-8")
    h = hashlib.sha256(seed_bytes + key_bytes).digest()
    val = struct.unpack(">I", h[:4])[0]
    return val % 10000


# ─── Operator Evaluation ─────────────────────────────────────────────────────

_regex_cache: Dict[str, re.Pattern[str]] = {}


def _evaluate_operator(op: str, attr_val: str, values: List[str]) -> bool:
    if op == "eq":
        return len(values) > 0 and attr_val == values[0]
    elif op == "neq":
        return len(values) > 0 and attr_val != values[0]
    elif op == "in":
        return attr_val in values
    elif op == "not_in":
        return attr_val not in values
    elif op == "contains":
        return len(values) > 0 and values[0] in attr_val
    elif op == "starts_with":
        return len(values) > 0 and attr_val.startswith(values[0])
    elif op == "ends_with":
        return len(values) > 0 and attr_val.endswith(values[0])
    elif op == "gt":
        return _compare_numeric(attr_val, values) > 0
    elif op == "lt":
        return _compare_numeric(attr_val, values) < 0
    elif op == "regex":
        if len(values) == 0:
            return False
        pattern = values[0]
        if pattern not in _regex_cache:
            try:
                _regex_cache[pattern] = re.compile(pattern)
            except re.error:
                return False
        return bool(_regex_cache[pattern].search(attr_val))
    else:
        return False


def _compare_numeric(attr_val: str, values: List[str]) -> int:
    if len(values) == 0:
        return 0
    try:
        a = float(attr_val)
    except (ValueError, TypeError):
        a = 0.0
    try:
        b = float(values[0])
    except (ValueError, TypeError):
        b = 0.0
    if a > b:
        return 1
    if a < b:
        return -1
    return 0


# ─── Attribute Resolution ────────────────────────────────────────────────────


def _get_attribute_value(attribute: str, ctx: EvalContext) -> str:
    if attribute == "key":
        return ctx.key
    val = ctx.attributes.get(attribute)
    if val is not None:
        return str(val)
    return ""


# ─── Clause & Rule Matching ──────────────────────────────────────────────────


def _matches_clause(clause: Clause, ctx: EvalContext) -> bool:
    attr_val = _get_attribute_value(clause.attribute, ctx)
    result = _evaluate_operator(clause.operator, attr_val, clause.values)
    return (not result) if clause.negate else result


def _matches_rule(rule: TargetingRule, ctx: EvalContext) -> bool:
    return all(_matches_clause(c, ctx) for c in rule.clauses)


# ─── Rollout Resolution ──────────────────────────────────────────────────────


def _rollout_variation(rc: RolloutConfig, ctx: EvalContext) -> Any:
    if not rc.variations:
        return None
    if len(rc.variations) == 1:
        return rc.variations[0].variation

    bucket_key = ctx.key
    if rc.bucket_by and rc.bucket_by in ctx.attributes:
        bucket_key = str(ctx.attributes[rc.bucket_by])

    bucket = hash_bucket(bucket_key, rc.seed)

    cumulative = 0
    for wv in rc.variations:
        cumulative += wv.weight
        if bucket < cumulative:
            return wv.variation

    return rc.variations[-1].variation


def _resolve_variation(rule: TargetingRule, ctx: EvalContext) -> Any:
    if rule.rollout is not None:
        return _rollout_variation(rule.rollout, ctx)
    return rule.variation


def _resolve_fallthrough(ft: RolloutConfig, ctx: EvalContext) -> Any:
    return _rollout_variation(ft, ctx)


# ─── Flag Store ───────────────────────────────────────────────────────────────


class FlagStore:
    """In-memory store for flag definitions and their environment configs."""

    def __init__(self) -> None:
        self.flags: Dict[str, Flag] = {}  # flag_key -> Flag
        self.flag_environments: Dict[str, FlagEnvironment] = {}  # flag_id -> FlagEnvironment

    def update(self, flags: List[Flag], flag_envs: List[FlagEnvironment]) -> None:
        self.flags.clear()
        for f in flags:
            self.flags[f.key] = f
        self.flag_environments.clear()
        for fe in flag_envs:
            self.flag_environments[fe.flag_id] = fe


# ─── Evaluator ────────────────────────────────────────────────────────────────


class Evaluator:
    """Local flag evaluation engine mirroring the Go Evaluator."""

    def __init__(self, store: FlagStore) -> None:
        self._store = store

    def evaluate(
        self, flag_key: str, ctx: EvalContext, default_value: Any = None
    ) -> EvalResult:
        """Evaluate a single flag for the given context."""
        now = datetime.now(timezone.utc).isoformat()

        flag = self._store.flags.get(flag_key)
        if flag is None:
            return EvalResult(
                flag_key=flag_key,
                value=default_value,
                reason="ERROR",
                version=0,
                timestamp=now,
            )

        flag_env = self._store.flag_environments.get(flag.id)
        if flag_env is None:
            return EvalResult(
                flag_key=flag_key,
                value=default_value,
                reason="ERROR",
                version=0,
                timestamp=now,
            )

        # Kill switch
        if flag.kill_switch:
            return EvalResult(
                flag_key=flag_key,
                value=flag_env.off_variation if flag_env.off_variation is not None else default_value,
                reason="KILL_SWITCH",
                version=flag_env.version,
                timestamp=now,
            )

        # Disabled
        if not flag.enabled or not flag_env.enabled:
            return EvalResult(
                flag_key=flag_key,
                value=flag_env.off_variation if flag_env.off_variation is not None else default_value,
                reason="OFF",
                version=flag_env.version,
                timestamp=now,
            )

        # Dependency checks
        if flag.depends_on:
            for dep in flag.depends_on:
                dep_result = self.evaluate(dep, ctx, None)
                if dep_result.reason in ("OFF", "ERROR"):
                    return EvalResult(
                        flag_key=flag_key,
                        value=flag_env.off_variation if flag_env.off_variation is not None else default_value,
                        reason="DEPENDENCY_FAILED",
                        version=flag_env.version,
                        timestamp=now,
                    )

        # Targeting rules
        for rule in flag_env.rules:
            if _matches_rule(rule, ctx):
                value = _resolve_variation(rule, ctx)
                return EvalResult(
                    flag_key=flag_key,
                    value=value,
                    reason="RULE_MATCH",
                    version=flag_env.version,
                    timestamp=now,
                )

        # Fallthrough
        value = _resolve_fallthrough(flag_env.fallthrough, ctx)
        return EvalResult(
            flag_key=flag_key,
            value=value,
            reason="FALLTHROUGH",
            version=flag_env.version,
            timestamp=now,
        )

    def evaluate_all(self, ctx: EvalContext) -> List[EvalResult]:
        """Evaluate all flags in the store for the given context."""
        results: List[EvalResult] = []
        for flag_key in self._store.flags:
            results.append(self.evaluate(flag_key, ctx))
        return results

File: internal/flag/evaluator.go
package flag

import (
	"crypto/sha256"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"github.com/Sriniketh24/rollout/internal/models"
)

type FlagStore interface {
	GetFlag(projectID, flagKey string) (*models.Flag, error)
	GetFlagEnvironment(flagID, envID string) (*models.FlagEnvironment, error)
	GetAllFlagsForEnvironment(projectID, envID string) ([]models.Flag, []models.FlagEnvironment, error)
}

type Evaluator struct {
	store     FlagStore
	cache     sync.Map
	cacheTTL  time.Duration
}

func NewEvaluator(store FlagStore, cacheTTL time.Duration) *Evaluator {
	return &Evaluator{
		store:    store,
		cacheTTL: cacheTTL,
	}
}

func (e *Evaluator) Evaluate(projectID, envID, flagKey string, ctx models.EvalContext) models.EvalResult {
	now := time.Now()

	flag, flagEnv, err := e.resolveFlag(projectID, envID, flagKey)
	if err != nil {
		return models.EvalResult{
			FlagKey:   flagKey,
			Value:     json.RawMessage(`null`),
			Reason:    models.ReasonError,
			Timestamp: now,
		}
	}

	if flag.KillSwitch {
		return models.EvalResult{
			FlagKey:   flagKey,
			Value:     flagEnv.OffVariation,
			Reason:    models.ReasonKillSwitch,
			Version:   flagEnv.Version,
			Timestamp: now,
		}
	}

	if !flag.Enabled || !flagEnv.Enabled {
		return models.EvalResult{
			FlagKey:   flagKey,
			Value:     flagEnv.OffVariation,
			Reason:    models.ReasonOff,
			Version:   flagEnv.Version,
			Timestamp: now,
		}
	}

	if len(flag.DependsOn) > 0 {
		for _, dep := range flag.DependsOn {
			depResult := e.Evaluate(projectID, envID, dep, ctx)
			if depResult.Reason == models.ReasonOff || depResult.Reason == models.ReasonError {
				return models.EvalResult{
					FlagKey:   flagKey,
					Value:     flagEnv.OffVariation,
					Reason:    models.ReasonDependency,
					Version:   flagEnv.Version,
					Timestamp: now,
				}
			}
		}
	}

	for _, rule := range flagEnv.Rules {
		if matchesRule(rule, ctx) {
			value := resolveVariation(rule, ctx)
			return models.EvalResult{
				FlagKey:   flagKey,
				Value:     value,
				Reason:    models.ReasonRuleMatch,
				Version:   flagEnv.Version,
				Timestamp: now,
			}
		}
	}

	value := resolveFallthrough(flagEnv.Fallthrough, ctx)
	return models.EvalResult{
		FlagKey:   flagKey,
		Value:     value,
		Reason:    models.ReasonFallthrough,
		Version:   flagEnv.Version,
		Timestamp: now,
	}
}

func (e *Evaluator) EvaluateAll(projectID, envID string, ctx models.EvalContext) []models.EvalResult {
	flags, flagEnvs, err := e.store.GetAllFlagsForEnvironment(projectID, envID)
	if err != nil {
		return nil
	}

	envMap := make(map[string]models.FlagEnvironment)
	for _, fe := range flagEnvs {
		envMap[fe.FlagID] = fe
	}

	results := make([]models.EvalResult, 0, len(flags))
	for _, f := range flags {
		result := e.Evaluate(projectID, envID, f.Key, ctx)
		results = append(results, result)
	}
	return results
}

func (e *Evaluator) resolveFlag(projectID, envID, flagKey string) (*models.Flag, *models.FlagEnvironment, error) {
	flag, err := e.store.GetFlag(projectID, flagKey)
	if err != nil {
		return nil, nil, err
	}

	flagEnv, err := e.store.GetFlagEnvironment(flag.ID, envID)
	if err != nil {
		return nil, nil, err
	}

	return flag, flagEnv, nil
}

func (e *Evaluator) InvalidateCache(flagKey string) {
	e.cache.Delete(flagKey)
}

func matchesRule(rule models.TargetingRule, ctx models.EvalContext) bool {
	for _, clause := range rule.Clauses {
		if !matchesClause(clause, ctx) {
			return false
		}
	}
	return true
}

func matchesClause(clause models.Clause, ctx models.EvalContext) bool {
	attrVal := getAttributeValue(clause.Attribute, ctx)
	result := evaluateOperator(clause.Operator, attrVal, clause.Values)
	if clause.Negate {
		return !result
	}
	return result
}

func getAttributeValue(attr string, ctx models.EvalContext) string {
	if attr == "key" {
		return ctx.Key
	}
	if v, ok := ctx.Attributes[attr]; ok {
		return fmt.Sprintf("%v", v)
	}
	return ""
}

func evaluateOperator(op models.Operator, attrVal string, values []string) bool {
	switch op {
	case models.OpEquals:
		return len(values) > 0 && attrVal == values[0]
	case models.OpNotEquals:
		return len(values) > 0 && attrVal != values[0]
	case models.OpIn:
		for _, v := range values {
			if attrVal == v {
				return true
			}
		}
		return false
	case models.OpNotIn:
		for _, v := range values {
			if attrVal == v {
				return false
			}
		}
		return true
	case models.OpContains:
		if len(values) == 0 {
			return false
		}
		return containsStr(attrVal, values[0])
	case models.OpStartsWith:
		if len(values) == 0 {
			return false
		}
		return len(attrVal) >= len(values[0]) && attrVal[:len(values[0])] == values[0]
	case models.OpEndsWith:
		if len(values) == 0 {
			return false
		}
		return len(attrVal) >= len(values[0]) && attrVal[len(attrVal)-len(values[0]):] == values[0]
	case models.OpGreaterThan:
		return compareNumeric(attrVal, values) > 0
	case models.OpLessThan:
		return compareNumeric(attrVal, values) < 0
	default:
		return false
	}
}

func containsStr(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

func compareNumeric(attrVal string, values []string) int {
	if len(values) == 0 {
		return 0
	}
	var a, b float64
	fmt.Sscanf(attrVal, "%f", &a)
	fmt.Sscanf(values[0], "%f", &b)
	if a > b {
		return 1
	}
	if a < b {
		return -1
	}
	return 0
}

func resolveVariation(rule models.TargetingRule, ctx models.EvalContext) json.RawMessage {
	if rule.Rollout != nil {
		return rolloutVariation(*rule.Rollout, ctx)
	}
	return rule.Variation
}

func resolveFallthrough(ft models.RolloutConfig, ctx models.EvalContext) json.RawMessage {
	return rolloutVariation(ft, ctx)
}

func rolloutVariation(rc models.RolloutConfig, ctx models.EvalContext) json.RawMessage {
	if len(rc.Variations) == 0 {
		return json.RawMessage(`null`)
	}
	if len(rc.Variations) == 1 {
		return rc.Variations[0].Variation
	}

	bucketKey := ctx.Key
	if rc.BucketBy != "" {
		if v, ok := ctx.Attributes[rc.BucketBy]; ok {
			bucketKey = fmt.Sprintf("%v", v)
		}
	}

	bucket := hashBucket(bucketKey, rc.Seed)

	cumulative := 0
	for _, wv := range rc.Variations {
		cumulative += wv.Weight
		if bucket < cumulative {
			return wv.Variation
		}
	}

	return rc.Variations[len(rc.Variations)-1].Variation
}

// hashBucket produces a deterministic bucket in [0, 10000) for sticky bucketing.
func hashBucket(key string, seed int64) int {
	h := sha256.New()
	seedBytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(seedBytes, uint64(seed))
	h.Write(seedBytes)
	h.Write([]byte(key))
	sum := h.Sum(nil)
	val := binary.BigEndian.Uint32(sum[:4])
	return int(val % 10000)
}
